pkg/frontend/transport/handler.go

```diff
@@ -440,13 +442,31 @@ func writeError(w http.ResponseWriter, err error) {
 		}
 	}
 
-	// if the error is an APIError, ensure it gets written as a JSON response
-	if resp, ok := apierror.HTTPResponseFromError(err); ok {
-		_ = httpgrpc.WriteResponse(w, resp)
+	var (
+		res *httpgrpc.HTTPResponse
+		ok  bool
+	)
+
+	// If the error is an APIError, ensure it gets written as a JSON response.
+	// Otherwise, check if there's a response encoded in the gRPC error.
+	res, ok = apierror.HTTPResponseFromError(err)
+	if !ok {
+		res, ok = httpgrpc.HTTPResponseFromError(err)
+	}
+
+	// If we've been able to get the HTTP response from the error, then we send
+	// it with the right status code and response body content.
+	if res != nil {
+		statusCode = int(res.Code)
+		_ = httpgrpc.WriteResponse(w, res)
 		return
 	}
 
-	httpgrpc.WriteError(w, err)
+	// Otherwise, we do fallback to a 5xx error, returning the non-formatted error
+	// message in the response body.
+	statusCode = http.StatusInternalServerError
```

Comment:
We have a utility function that can extract the grpc status code: https://github.com/grafana/dskit/blob/4beef469a18ec34fdc12d1fd678b1f0c018de4f4/grpcutil/status.go#L48 . As far as I understand, those are compatible with HTTP status codes, so might make sense to use it

Reply:
I don't understand the comment in this line. This is the case where we haven't been able to detect the HTTP status code from the error, so we return a generic 500 (the previous code did the exact same, look at `httpgrpc.WriteError()` implementation).

Reply:
I remembered that in some interactions with downstream we use GRPC and that has different error handling, but I'm probably mixing things up. Ignore it.